luarules/gadgets/unit_dgun_behaviour.lua

```diff
@@ -57,63 +82,119 @@ local function addVolumetricDamage(projectileID)
 		-- place projectile slightly under ground to ensure fiery trail
 		local verticalOffset = 1
-		Spring.SetProjectilePosition(proID, x, math.max(Spring.GetGroundHeight(x, z), 0) - verticalOffset, z)
+		spSetProjectilePosition(proID, x, math.max(spGetGroundHeight(x, z), 0) - verticalOffset, z)
 
 		-- NB: no removal; do this every frame so that it doesn't fly off a cliff or something
 	end
+
+	-- Without defining a time to live (TTL) for the DGun, it will live forever until it reaches maximum range. This means it would deal infinite damage to shields until it depleted them.
+	if next(dgunTimeouts) == nil then
+		frameCounter = 0
+	else
+		frameCounter = frameCounter + 1
+
+		for proID, timeout in pairs(dgunTimeouts) do
+			if frameCounter > timeout then
+				spDeleteProjectile(proID)
+				flyingDGuns[proID] = nil
+				groundedDGuns[proID] = nil
+				dgunTimeouts[proID] = nil
+			end
+		end
+	end
 end
 
 function gadget:UnitPreDamaged(unitID, unitDefID, unitTeam, damage, paralyzer, weaponDefID, projectileID, attackerID, attackerDefID, attackerTeam)
 	if dgunWeapons[weaponDefID] and isCommander[attackerDefID] and (isCommander[unitDefID] or isDecoyCommander[unitDefID]) then
 		if isDecoyCommander[unitDefID] then
 			return dgunWeapons[weaponDefID].damages[0]
 		else
-			Spring.DeleteProjectile(projectileID)
-			local x, y, z = Spring.GetUnitPosition(unitID)
-			Spring.SpawnCEG("dgun-deflect", x, y, z, 0, 0, 0, 0, 0)
+			spDeleteProjectile(projectileID)
+			local x, y, z = spGetUnitPosition(unitID)
+			spSpawnCEG("dgun-deflect", x, y, z, 0, 0, 0, 0, 0)
 			local armorClass = UnitDefs[unitDefID].armorType
 			return dgunWeapons[weaponDefID].damages[armorClass]
 		end
 	end
+
 	return damage
 end
+
+local lastShieldFrameCheck = {}
+
+function gadget:ShieldPreDamaged(proID, proOwnerID, shieldEmitterWeaponNum, shieldCarrierUnitID, bounceProjectile, beamEmitterWeaponNum, beamEmitterUnitID, startX, startY, startZ, hitX, hitY, hitZ)
+	if proID > -1 and dgunTimeouts[proID] then
+		local proDefID = spGetProjectileDefID(proID)
+		local shieldEnabledState, shieldPower = spGetUnitShieldState(shieldCarrierUnitID)
+		local damage = WeaponDefs[proDefID].damages[11] or WeaponDefs[proDefID].damages[2]
+
+		if modOptions.shieldsrework == false and hitX > 0 and lastShieldFrameCheck[shieldCarrierUnitID] ~= frameCounter then
+			shieldPower = math.max(shieldPower - damage, 0)
+			spSetUnitShieldState(shieldCarrierUnitID, shieldEmitterWeaponNum, shieldEnabledState, shieldPower)
+			lastShieldFrameCheck[shieldCarrierUnitID] = frameCounter
+		end
+
+		local originX, originY, originZ = unpack(dgunOrigins[proID])
+
+		if shieldPower > 100 then
+			local dirX = hitX - originX
+			local dirZ = hitZ - originZ
+			local length = math.sqrt(dirX * dirX + dirZ * dirZ)
+			dirX = dirX / length
+			dirZ = dirZ / length
+			local newX = hitX - (WeaponDefs[proDefID].projectilespeed * 4) * dirX
+			local newZ = hitZ - (WeaponDefs[proDefID].projectilespeed * 4) * dirZ
```

Comment:
Why does it need to be proportional to speed? Shouldn't a static value work, since regardless of speed, the projectile will stall on the shield until it's depleted?

Reply:
It would work if you could gauruntee all DGun projectiles will move at the same speed of... 300 ish?

The scavenger boss dgun moves faster than that and might exhibit anomalous behavior with a shared static dgun projectile setback distance.